rfcs/text/0014_notification_service.md

```diff
@@ -0,0 +1,373 @@
+Filtering allows users to hide notifications they aren’t interested in. Kibana UI attaches a list of filters when requesting notifications.
+Kibana stores filter as a part of User settings data to provide a seamless experience between different deployments on Cloud and user sessions.
+
+
+# Drawbacks
+
+- It is not entirely clear what functionality will be supported by UNS and when this will be known.
+It forces us to evaluate each feature separately as to how difficult it will be to adapt it to UNS later.
+- The service relies on the yet-to-be-added ability to uniquely identify the user, enumerate users with a given set of permissions.
+- The service doesn't provide a built-in security mechanism but relies on a notification source implementation.
+- Lack of a centralized i18n service makes message transferring over-bloated due to the necessity to send all the possible translations over the wire.
+
+
+# Alternatives
+
+### Threat Cloud as another pluggable storage mechanism.
+Kibana notification service collects internal notifications and shows them in the UI for both on-prem and Cloud environments.
+Kibana reuses Notification service implementation on Cloud but enhances it with additional pluggable storage of external notifications.
+![img](../images/0014/alternative_implementation.png)
+
+The main benefit of this approach is that Kibana always uses the same service in any environment.
+But in addition to the restrictions listed in the [Drawbacks](#drawbacks) section, the following are added:
+- Each retrieval of a notification list still requires waiting for a roundtrip to UNS.
+- The list of Kibana notifications stored in an instance might be out of sync with a copy stored in UNS due to network problems. 
+- Storing notifications in the Cloud deployment and their copy in UNS consumes twice the disk space.
+
+# Adoption strategy
+
+We start with introducing Kibana Notification Service as a Kibana plugin, adding more notification sources along the road.
+Integration with UNS is something that is not going to happen any time soon. The work can be divided into the following phases:
+
+### Phase I
+Kibana shows notifications created by the Newsfeed Kibana plugin in Notification flyout.
+![img](../images/0014/kns_flyout.png)
+A notification cannot have a particular recipient in the lack of a way to identify a user in the system. A notification will be shown to all the users instead.
+User-specific notification state (*read_status*) is not stored in Kibana but browser Local Storage.
+### Phase II
+When [user profiles](https://github.com/elastic/kibana/issues/17888) are supported, Kibana UI starts showing user-specific notifications.
+KIbana UI supports applying filters to the notification list.
+The very first three candidates for the integration with Notification systems:
+- Data plugin - Search sessions
+- Security solutions plugin - Cases
+- Alerting plugin - Alerts
+
+The notification system supports a user-specific state for notifications.
```

Comment:
If it's not already done for Phase I, then it sounds like Phase II would need space-awareness? Or will you always see notifications from all of your spaces?

Reply:
As far as I'm concerned, notifications are space-agnostic. 

Spaces are difficult to use on the Cloud because a user does not have an active Space during the session in Cloud Console UI.

For on-prem Kibana, it makes sense to show notifications from all the user's Spaces. Say, you have an alert triggered in `Space-A`, why shouldn't you receive a notification even if your active Space is `Space-B`? 

@alexfrancoeur could you please confirm these conclusions?

Reply:
Confirming the current notification UX/UI is space agnostic, and the spirit of the feature is to provide visibility across our stack/products where ever you are (as you illustrate @restrry ).

That being said, I could see filtering by space making sense at some point. If we wanted to include the concept of a space as an optional part of the notification data model (intended as a way to organize notifications) what challenges would we run into?

Reply:
> If we wanted to include the concept of a space as an optional part of the notification data model (intended as a way to organize notifications) what challenges would we run into?

Let's continue with Alerting example. When an alert notification is created, it stores in what Space the alert has been created. https://github.com/elastic/kibana/blob/master/x-pack/plugins/alerts/server/alerts_client/alerts_client.ts#L1335
It can pass `spaceId` as a part of Notification content.

Reply:
> It can pass `spaceId` as a part of Notification content. Notification service itself cannot get this information in runtime.

Agreed that the default experience should be showing all notifications across all spaces, but if we wanted to support filtering by space, couldn't we automatically source the current space from the request context when a notification is created? There may be some licensing and/or technical limitations that make this difficult, but it may be worth solving those if we want this experience.

Reply:
> couldn't we automatically source the current space from the request context when a notification is created?

Sure, we can do it when we have a request context (as in the case of Background search). but it's not always the case - we don't have request context when an alert is triggered. I'd consider `SpaceId` to be passed to provide A-Single-Way-To-Do-Things™.